New features to THRIFT and to pySTEL/vmec

ajchcoelho opened the pull request:
This merge brings the following new features to `THRIFT`:
- we can now provide non-constant heat diffusivities via the input h5 file. This allows for more realistic transport simulations;
- the calls to `VMEC` are done using as input `dI/ds` computed directly from `THRIFT_I` and not through `THRIFT_J`. This makes the iota profiles smoother in some situations.

The merge brings some novelties to `pySTEL` too, the most important ones being on `vmec.py`: 
-  a routine that calculates the covariant tensor `g_ij` has been created;
- the routine that gives `<|nabla_rho|^2 >` has been fixed;
- a routine that computes `<|nabla_rho|>` has been added.

ajchcoelho committed:
Merge branch 'feature/plasma_solver_faster' into feature/external_heat_diffusivities_THRIFT

ajchcoelho committed:
pySTEL: Add possibility to read external chi in THRIFT transport solver

The external chi profiles should be given as function of rho and a/LT and are normalized to chi_gB

ajchcoelho committed:
pySTEL: frac_alpha_heating is now an input parameter

Before it was fixed as: 80% to electrons, 10% to deuterium and 10% to tritium. Now is given as an array in the input namelist

ajchcoelho committed:
pySTEL: Correct comment

ajchcoelho committed:
pySTEL: make tau_fast_alphas and alpha_chi_external input variables

ajchcoelho committed:
THRIFT: Remove deprecated boolean variable

ajchcoelho committed:
THRIFT: Save Dn_total and cn_total

ajchcoelho committed:
pySTEL: Add functions to thrift class and improve others

ajchcoelho committed:
pySTEL: Add functions to vmec class that return (R,Z) and triangulations

ajchcoelho committed:
THRIFT: Save dV/dr and energy sources in output plasma solver file

ajchcoelho committed:
THRIFT: Save add_NEO boolean in plasma solver output h5 file

ajchcoelho committed:
pySTEL: Add covariant tensor computation, fix gradrho2 and add gradrho

The function calc_grad_rhosq was broken. Now it is fixed. Also added calc_grad_rho. A new routine that returns the covariant tensor elements of the VMEC equilibrium has been included

ajchcoelho committed:
THRIFT: Feed VMEC with dI/ds from THRIFT_I and not through THRIFT_J

Taking the EZSpline derivative of THRIFT_I and feeding it to VMEC gives smoother iota profiles than feeding VMEC with central finite difference of THRIFT_J in some situations

ajchcoelho committed:
pySTEL: Improve thrift class

lazersos commented on pySTEL/libstell/vmec.py:
```diff
@@ -199,58 +199,214 @@ def calc_magwell(self):
+		dRds = np.gradient(R,s, axis=0)
+		dZds = np.gradient(Z,s, axis=0)
+		#
+		dRdu = np.gradient(R, theta.ravel(), axis=1)
```
dX/du and dV/du should be calculated from the derivatives of the Fourier Harmonics for a more accurate representation.

ajchcoelho replied:
Done. Now the poloidal and toroidal derivatives are computed via the Fourier Harmonics and not through np.gradient. 

I have also changed the way the radial derivatives, d/ds, are performed, by mimicking what vmec does internally. Now even m-modes and odd m-modes are treated differently, the former interpolated as `s` and the latter as `sqrt(s)`. This overcomes the issue of the derivatives diverging close to the axis.

An interesting test to the newly implemented `calc_covariant_tensor` function in `vmec.py`  is to compare the jacobian of an wout file with that computed through the determinant of the g_cov matrix returned by  `calc_covariant_tensor` . They agree:

<img width="1296" height="888" alt="image" src="https://github.com/user-attachments/assets/cb8057b8-0465-463c-b9ae-758e805629a9" />

lazersos reviewed (changes requested):
There's a minor point here about using the derivatives of the functions and not `gradient` to compute dX/du and dX/dv.

ajchcoelho committed:
pySTEL: bugfix in plasma class

ajchcoelho committed:
pySTEL: bugfix in thrift class

ajchcoelho committed:
pySTEL: Compute derivatives as in vmec_utils.f.

s-derivatives are computed using different weights for odd and even m-modes. theta and phi derivatives are now computed with the exact Fourier harmonics and not with np.gradient

ajchcoelho committed:
THRIFT: bugfix in eq_phiedge

Taking the absolute value of eq_phiedge is not correct. We must keep the sign

ajchcoelho commented:
I have also introduced a bugfix to THRIFT. When PHI_EDGE<0, `THRIFT` was not caring about its sign, which it turns out to be quite important.

lazersos reviewed (approved):
Looks good now.
